src/main/java/com/mongodb/jdbc/oidc/RFC8252HttpServer.java

```diff
@@ -0,0 +1,256 @@
+    private static final String ERROR_URI_KEY = "error_uri";
+    private static final String ERROR_KEY = "error";
+    private static final String ERROR_DESCRIPTION_KEY = "error_description";
+
+    // server endpoints
+    private static final String ACCEPTED_ENDPOINT = "/accepted";
+    private static final String CALLBACK_ENDPOINT = "/callback";
+    private static final String REDIRECT_ENDPOINT = "/redirect";
+
+    private HttpServer server;
+    private final TemplateEngine templateEngine;
+    private final BlockingQueue<OidcResponse> oidcResponseQueue;
+
+    public RFC8252HttpServer() {
+        templateEngine = createTemplateEngine();
+        oidcResponseQueue = new LinkedBlockingQueue<>();
+    }
+
+    /**
+     * Starts the HTTP server and sets up the necessary contexts and handlers.
+     *
+     * @throws IOException if an I/O error occurs while creating or starting the server
+     */
+    public void start() throws IOException {
+        server = HttpServer.create(new InetSocketAddress(DEFAULT_REDIRECT_PORT), 0);
+
+        server.createContext(CALLBACK_ENDPOINT, new CallbackHandler());
+        server.createContext(REDIRECT_ENDPOINT, new CallbackHandler());
+        server.createContext(ACCEPTED_ENDPOINT, new AcceptedHandler());
+        server.setExecutor(null);
+        server.start();
+    }
+
+    /**
+     * Blocks until an OIDC response is available in the queue.
+     *
+     * @return the OIDC response
+     * @throws InterruptedException if the current thread is interrupted while waiting
+     */
+    public OidcResponse getOidcResponse() throws InterruptedException {
+        return oidcResponseQueue.take();
+    }
+
+    public void stop() {
+        if (server != null) {
+            server.stop(0);
+        }
+    }
+
+    /**
+     * Creates and configures the template engine.
+     *
+     * @return the configured template engine
+     */
+    private TemplateEngine createTemplateEngine() {
+        TemplateEngine templateEngine = new TemplateEngine();
+        ClassLoaderTemplateResolver templateResolver = new ClassLoaderTemplateResolver();
+        templateResolver.setPrefix("/templates/");
+        templateResolver.setSuffix(".html");
+        templateEngine.setTemplateResolver(templateResolver);
+        return templateEngine;
+    }
+
+    /** HTTP handler for handling the callback and redirect endpoints. */
+    private class CallbackHandler implements HttpHandler {
+
+        @Override
+        public void handle(HttpExchange exchange) throws IOException {
+            Map<String, String> queryParams = parseQueryParams(exchange);
+            OidcResponse oidcResponse = new OidcResponse();
+
+            if (queryParams.containsKey(CODE)) {
+                oidcResponse.setCode(queryParams.get(CODE));
+                oidcResponse.setState(queryParams.getOrDefault(STATE, ""));
+                if (!putOidcResponse(exchange, oidcResponse)) {
+                    return;
```

Comment:
We should close the exchange or the http connection will stay open, isn't it?

Reply:
The `sendResponse()` will close it in the false case for `putOidcResponse()`.